File: src/client/python/agentguard/guard.py
"""AgentGuard: the public client facade."""
from __future__ import annotations

from pathlib import Path
from typing import Any, Callable

from agentguard.adapters.agent import default_agent_adapters, select_agent_adapter
from agentguard.adapters.llm import default_llm_adapters, select_llm_adapter
from agentguard.audit.logger import AuditLogger
from agentguard.audit.recorder import AuditRecorder
from agentguard.checkers.manager import CheckerManager
from agentguard.harness.event_bus import EventBus
from agentguard.harness.lifecycle import Lifecycle
from agentguard.harness.runtime import HarnessRuntime
from agentguard.plugins.builtin.agentdog_proxy import AgentDoGProxyPlugin
from agentguard.plugins.manager import PluginManager
from agentguard.rules.loader import load_policy
from agentguard.sandbox.executor import SandboxExecutor
from agentguard.schemas.context import RuntimeContext
from agentguard.skill_client.registry_proxy import SkillRegistryProxy
from agentguard.skill_client.remote_runner import RemoteSkillRunner
from agentguard.tools.degrade import ToolDegradeManager
from agentguard.tools.metadata import ToolMetadata
from agentguard.tools.registry import ToolRegistry
from agentguard.tools.wrapper import ToolWrapper
from agentguard.u_guard.decision_cache import DecisionCache
from agentguard.u_guard.enforcer import UGuardEnforcer
from agentguard.u_guard.policy_snapshot import PolicySnapshot
from agentguard.u_guard.remote_client import RemoteGuardClient


class AgentGuard:
    """Lightweight client-side Harness/U-Guard runtime."""

    def __init__(
        self,
        session_id: str,
        *,
        user_id: str | None = None,
        agent_id: str | None = None,
        policy: str | None = None,
        server_url: str | None = None,
        api_key: str | None = None,
        environment: str | None = None,
        sandbox: str = "local",
        sandbox_profile: Any = None,
        enable_agentdog: bool = False,
        max_steps: int = 12,
        max_tool_calls: int = 24,
        window_size: int = 8,
        audit_path: str | None = None,
        remote_timeout_s: float = 5.0,
        remote_retries: int = 2,
    ) -> None:
        snapshot = self._load_snapshot(policy)
        self.context = RuntimeContext(
            session_id=session_id,
            user_id=user_id,
            agent_id=agent_id,
            policy=policy,
            policy_version=snapshot.version,
            environment=environment,
        )

        self._remote = RemoteGuardClient(
            server_url,
            api_key=api_key,
            timeout_s=remote_timeout_s,
            retries=remote_retries,
        )
        self._cache = DecisionCache()
        self._enforcer = UGuardEnforcer(
            snapshot=snapshot,
            remote=self._remote,
            checker_manager=CheckerManager(),
            cache=self._cache,
        )
        self._sandbox = SandboxExecutor(sandbox, sandbox_profile)
        self._audit = AuditRecorder(session_id, AuditLogger(audit_path))
        self._registry = ToolRegistry()
        self._degrade = ToolDegradeManager()
        self._lifecycle = Lifecycle()
        self._bus = EventBus()
        self._plugins = PluginManager(self._lifecycle)

        self.runtime = HarnessRuntime(
            context=self.context,
            enforcer=self._enforcer,
            sandbox=self._sandbox,
            audit=self._audit,
            registry=self._registry,
            degrade_manager=self._degrade,
            lifecycle=self._lifecycle,
            event_bus=self._bus,
            max_steps=max_steps,
            max_tool_calls=max_tool_calls,
            window_size=window_size,
        )

        self._agent_adapters = default_agent_adapters()
        self._llm_adapters = default_llm_adapters()
        self._skills = SkillRegistryProxy(
            remote=RemoteSkillRunner(server_url, api_key=api_key) if server_url else None
        )

        if enable_agentdog:
            self.register_plugin(AgentDoGProxyPlugin())
        self._plugins.start_session(self.context)

    # ---- policy --------------------------------------------------------
    @staticmethod
    def _load_snapshot(policy: str | None) -> PolicySnapshot:
        rules = None
        if policy:
            for cand in (policy, f"rules/examples/{policy}.json", f"rules/{policy}.json"):
                if cand and Path(cand).exists():
                    rules = load_policy(cand)
                    break
        if rules is None:
            rules = load_policy(None)
        return PolicySnapshot(version=policy or "builtin", rules=rules)

    def load_policy_snapshot(self, snapshot: PolicySnapshot | dict[str, Any]) -> None:
        snap = snapshot if isinstance(snapshot, PolicySnapshot) else PolicySnapshot.from_dict(snapshot)
        self._enforcer.set_snapshot(snap)
        self.context.policy_version = snap.version

    # ---- wrapping ------------------------------------------------------
    def wrap_tool(self, fn: Callable[..., Any], **meta: Any) -> ToolWrapper:
        metadata = self.register_tool(fn, **meta)
        return ToolWrapper(fn, metadata, self.runtime)

    def wrap_agent(self, agent: Any) -> Any:
        adapter = select_agent_adapter(agent, self._agent_adapters)
        return adapter.wrap(agent, self.runtime)

    def wrap_llm(self, llm: Any) -> Any:
        adapter = select_llm_adapter(llm, self._llm_adapters)
        return adapter.wrap(llm, self.runtime)

    # ---- registration --------------------------------------------------
    def register_tool(self, fn: Callable[..., Any], **meta: Any) -> ToolMetadata:
        return self._registry.register(fn, **meta)

    def register_plugin(self, plugin: Any) -> Any:
        return self._plugins.register(plugin)

    def register_skill(self, skill: Any) -> Any:
        try:
            from skills.registry import get_registry  # noqa: PLC0415

            get_registry().register(skill)
        except Exception:
            pass
        return skill

    # ---- skills --------------------------------------------------------
    def run_skill(self, skill_name: str, input_data: dict[str, Any] | None = None) -> dict[str, Any]:
        return self._skills.run(skill_name, input_data or {})

    # ---- tools invocation (direct) ------------------------------------
    def invoke_tool(self, tool_name: str, arguments: dict[str, Any]) -> Any:
        reg = self._registry.get(tool_name)
        if reg is None:
            raise ValueError(f"tool not registered: {tool_name}")
        return self.runtime.invoke_tool(
            tool_name=tool_name, arguments=arguments, fn=reg.fn, metadata=reg.metadata
        )

    # ---- audit ---------------------------------------------------------
    def flush_audit(self) -> list[dict[str, Any]]:
        return self._audit.flush()

    @property
    def trace(self):
        return self.runtime.session.trace

    def close(self) -> None:
        self._plugins.end_session(self.runtime.session.trace, self.context)

File: src/client/js/agentguard/guard.js
"use strict";

const crypto = require("crypto");
const fs = require("fs");
const path = require("path");
const { defaultLLMAdapters, selectLLMAdapter } = require("./adapters/llm");
const { AuditLogger } = require("./audit/logger");
const { AuditRecorder } = require("./audit/recorder");
const { PluginManager } = require("./plugins/manager");
const { ClientConfigAPIServer } = require("./config_api");
const { EventBus } = require("./harness/event_bus");
const { Lifecycle } = require("./harness/lifecycle");
const { HarnessRuntime } = require("./harness/runtime");
const { loadPolicy } = require("./rules/loader");
const { SandboxExecutor } = require("./sandbox/executor");
const { RuntimeContext } = require("./schemas/context");
const { SkillRegistryProxy } = require("./skill_client/registry_proxy");
const { RemoteSkillRunner } = require("./skill_client/remote_runner");
const { ToolMetadata } = require("./tools/metadata");
const { ToolDegradeManager } = require("./tools/degrade");
const { ToolRegistry } = require("./tools/registry");
const { ToolWrapper } = require("./tools/wrapper");
const { UGuardEnforcer } = require("./u_guard/enforcer");
const { PolicySnapshot } = require("./u_guard/policy_snapshot");
const { RemoteGuardClient } = require("./u_guard/remote_client");
const { LangChainAgentAdapter } = require("./adapters/agent/langchain");
const { AutogenAgentAdapter } = require("./adapters/agent/autogen");
const { OpenAIAgentsAdapter } = require("./adapters/agent/openai_agents");

class AgentGuard {
  constructor(session_id, options = {}) {
    const pluginPayload = pluginConfigPayload(options.checker_config || options.checkerConfig || null);
    const snapshot = this.loadSnapshot(options.policy || null);
    this.session_key = options.session_key || options.sessionKey || generateSessionKey();
    this.context = new RuntimeContext({
      session_id,
      user_id: options.user_id || options.userId || null,
      agent_id: options.agent_id || options.agentId || null,
      policy: options.policy || null,
      policy_version: snapshot.version,
      environment: options.environment || null,
      metadata: {
        client_session_key: this.session_key,
        client_checker_config: pluginPayload,
        remote_checker_config: pluginPayload,
      },
    });
    this.remote = new RemoteGuardClient(options.server_url || options.serverUrl || null, {
      api_key: options.api_key || options.apiKey || null,
      session_id: this.context.session_id,
      agent_id: this.context.agent_id,
      user_id: this.context.user_id,
      session_key: this.session_key,
      timeout_s: options.remote_timeout_s ?? options.remoteTimeoutS ?? 5.0,
      retries: options.remote_retries ?? options.remoteRetries ?? 2,
    });
    this.enforcer = new UGuardEnforcer({
      snapshot,
      remote: this.remote,
      plugin_manager: new PluginManager({ config: options.checker_config || options.checkerConfig || null }),
    });
    this.sandbox = new SandboxExecutor(options.sandbox || "local", options.sandbox_profile || options.sandboxProfile || null);
    this.audit = new AuditRecorder(session_id, new AuditLogger(options.audit_path || options.auditPath || null));
    this.registry = new ToolRegistry();
    this.degrade = new ToolDegradeManager();
    this.lifecycle = new Lifecycle();
    this.bus = new EventBus();
    this.config_api = null;
    this.runtime = new HarnessRuntime({
      context: this.context,
      enforcer: this.enforcer,
      sandbox: this.sandbox,
      audit: this.audit,
      registry: this.registry,
      degrade_manager: this.degrade,
      lifecycle: this.lifecycle,
      event_bus: this.bus,
      max_steps: options.max_steps ?? options.maxSteps ?? 12,
      max_tool_calls: options.max_tool_calls ?? options.maxToolCalls ?? 24,
      window_size: options.window_size ?? options.windowSize ?? 8,
    });
    this.llm_adapters = defaultLLMAdapters();
    this.skills = new SkillRegistryProxy({
      remote: options.server_url || options.serverUrl
        ? new RemoteSkillRunner(options.server_url || options.serverUrl, {
            api_key: options.api_key || options.apiKey || null,
            session_id: this.context.session_id,
            agent_id: this.context.agent_id,
            user_id: this.context.user_id,
            session_key: this.session_key,
          })
        : null,
    });
    this.remote_session_registration = null;
    this.remote_session_registered = false;
    this.registerRemoteSession();
  }

  loadSnapshot(policy) {
    let rules = null;
    if (policy) {
      for (const candidate of [policy, path.join("rules", "examples", `${policy}.json`), path.join("rules", `${policy}.json`)]) {
        try {
          rules = loadPolicy(candidate);
          break;
        } catch (_) {
          continue;
        }
      }
    }
    if (!rules) {
      rules = loadPolicy(null);
    }
    return new PolicySnapshot({
      version: policy || "builtin",
      rules,
    });
  }

  load_policy_snapshot(snapshot) {
    const next = snapshot instanceof PolicySnapshot ? snapshot : PolicySnapshot.fromDict(snapshot);
    this.enforcer.set_snapshot(next);
    this.context.policy_version = next.version;
  }

  update_checker_config(checker_config, { sync_remote = true, syncRemote = sync_remote } = {}) {
    const payload = pluginConfigPayload(checker_config);
    this.context.metadata.client_checker_config = payload;
    this.enforcer.update_plugin_config(checker_config);
    if (syncRemote) {
      return this.syncRemoteSession();
    }
    return Promise.resolve();
  }

  async start_config_api({ host = "127.0.0.1", port = 38181, sync_remote = true, syncRemote = sync_remote } = {}) {
    const prevConfigUrl = this.context.metadata.client_config_url;
    const prevPluginListUrl = this.context.metadata.client_plugin_list_url;
    const prevHealthUrl = this.context.metadata.client_health_url;
    if (!this.config_api) {
      this.config_api = new ClientConfigAPIServer(this, { host, port });
    }
    this.context.metadata.client_config_url = this.config_api.plugin_config_url;
    this.context.metadata.client_plugin_list_url = this.config_api.plugin_list_url;
    this.context.metadata.client_health_url = this.config_api.health_url;
    const url = await this.config_api.start().catch(() => this.config_api.plugin_config_url);
    this.context.metadata.client_config_url = url;
    this.context.metadata.client_plugin_list_url = this.config_api.plugin_list_url;
    this.context.metadata.client_health_url = this.config_api.health_url;
    const urlsChanged = (
      prevConfigUrl !== url ||
      prevPluginListUrl !== this.config_api.plugin_list_url ||
      prevHealthUrl !== this.config_api.health_url
    );
    if (syncRemote && urlsChanged) {
      await this.syncRemoteSession();
    }
    return url;
  }

  async stop_config_api() {
    if (!this.config_api) {
      return;
    }
    await this.config_api.stop();
    this.config_api = null;
    delete this.context.metadata.client_config_url;
    delete this.context.metadata.client_plugin_list_url;
    delete this.context.metadata.client_health_url;
  }

  register_tool(fn, meta = {}) {
    const metadata = meta instanceof ToolMetadata
      ? this.registry.register(fn, meta)
      : this.registry.register(fn, null, meta);
    this.reportToolMetadata(metadata);
    return metadata;
  }

  wrap_tool(fn, meta = {}) {
    const metadata = this.register_tool(fn, meta);
    return new ToolWrapper(fn, metadata, this.runtime);
  }

  wrap_llm(llm) {
    const adapter = selectLLMAdapter(llm, this.llm_adapters);
    return adapter.wrap(llm, this.runtime);
  }

  attach_autogen(agent, options = {}) {
    return new AutogenAgentAdapter().attach(agent, this, options);
  }

  attach_langchain(agent, options = {}) {
    return new LangChainAgentAdapter().attach(agent, this, options);
  }

  attach_openai_agents(agent, options = {}) {
    return new OpenAIAgentsAdapter().attach(agent, this, options);
  }

  register_skill(skill) {
    try {
      const { getRegistry } = require("./skill_client");
      getRegistry().register(skill);
    } catch (_) {
      return skill;
    }
    return skill;
  }

  async run_skill(skill_name, input_data = {}) {
    return this.skills.run(skill_name, input_data);
  }

  async invoke_tool(tool_name, arguments_ = {}) {
    const registered = this.registry.get(tool_name);
    if (!registered) {
      throw new Error(`tool not registered: ${tool_name}`);
    }
    return this.runtime.invoke_tool({
      tool_name,
      arguments: arguments_,
      fn: registered.fn,
      metadata: registered.metadata,
    });
  }

  flush_audit() {
    return this.audit.flush();
  }

  get trace() {
    return this.runtime.session.trace;
  }

  async close() {
    await this.runtime.sync_local_cache_now({ reason: "session_close" });
    if (this.remote.enabled) {
      try {
        const registered = await this.ensureRemoteSessionRegistered();
        if (registered) {
          await this.remote.unregister_session();
          this.remote_session_registered = false;
          this.remote_session_registration = null;
        }
      } catch (_) {
        // swallow remote shutdown errors to match Python close()
      }
    }
    await this.stop_config_api();
  }

  ensureRemoteSessionRegistered() {
    if (!this.remote.enabled) {
      return Promise.resolve(false);
    }
    if (this.remote_session_registered) {
      return Promise.resolve(true);
    }
    if (this.remote_session_registration) {
      return this.remote_session_registration;
    }
    this.remote_session_registration = this.remote.register_session(this.context)
      .then(() => {
        this.remote_session_registered = true;
        this.remote_session_registration = null;
        return true;
      })
      .catch(() => {
        this.remote_session_registration = null;
        return false;
      });
    return this.remote_session_registration;
  }

  syncRemoteSession() {
    if (!this.remote.enabled) {
      return Promise.resolve(false);
    }
    return this.remote.register_session(this.context)
      .then(() => {
        this.remote_session_registered = true;
        return true;
      })
      .catch(() => false);
  }

  registerRemoteSession() {
    if (!this.remote.enabled) {
      return;
    }
    if (this.remote_session_registration) {
      return;
    }
    this.remote_session_registration = this.start_config_api({ port: 0, syncRemote: false })
      .catch(() => null)
      .then(() => this.remote.register_session(this.context))
      .then(() => {
        this.remote_session_registered = true;
        this.remote_session_registration = null;
        return true;
      })
      .catch(() => {
        this.remote_session_registration = null;
        return false;
      });
  }

  reportToolMetadata(metadata) {
    if (!this.remote.enabled) {
      return;
    }
    const toolPayload = {
      name: metadata.name,
      description: metadata.description,
      input_params: [...(metadata.required_args || [])],
      capabilities: [...(metadata.capabilities || [])],
      labels: {
        boundary: String((metadata.metadata || {}).boundary || "internal"),
        sensitivity: String((metadata.metadata || {}).sensitivity || "low"),
        integrity: String((metadata.metadata || {}).integrity || "trusted"),
        tags: [ ...(((metadata.metadata || {}).tags || metadata.capabilities || []).map((tag) => String(tag)).filter(Boolean)) ],
      },
    };
    this.ensureRemoteSessionRegistered()
      .then((registered) => (registered ? this.remote.report_tool(this.context, toolPayload) : null))
      .catch(() => {});
  }
}

function generateSessionKey() {
  return `sk-${crypto.randomBytes(32).toString("base64url")}`;
}

function pluginConfigPayload(checker_config) {
  if (checker_config == null) {
    return null;
  }
  if (typeof checker_config === "object") {
    return JSON.parse(JSON.stringify(checker_config));
  }
  const raw = fs.readFileSync(checker_config, "utf-8");
  const data = JSON.parse(raw);
  if (!data || typeof data !== "object" || Array.isArray(data)) {
    throw new Error("plugin config file must contain a JSON object");
  }
  return data;
}

module.exports = {
  AgentGuard,
};
